pkg/logging/logger.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2016-2019, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logging
+
+import (
+	"context"
+
+	"github.com/pulumi/pulumi/pkg/diag"
+	"github.com/pulumi/pulumi/pkg/resource"
+	"github.com/pulumi/pulumi/pkg/resource/provider"
+)
+
+// Logger wraps a time-ordered log set to allow batched logging of unique messages.
+type Logger struct {
+	messages *TimeOrderedLogSet
+	index    int
+	ctx      context.Context
+	host     *provider.HostClient
+	urn      resource.URN
+}
+
+// NewLogger returns an initialized Logger.
+func NewLogger(ctx context.Context, host *provider.HostClient, urn resource.URN) *Logger {
+	return &Logger{
+		messages: &TimeOrderedLogSet{},
+		ctx:      ctx,
+		host:     host,
+		urn:      urn,
+	}
+}
+
+// LogMessage adds a message to the log set and flushes the queue to the host.
+func (l *Logger) LogMessage(severity diag.Severity, s string) {
+	l.EnqueueMessage(severity, s)
+	l.LogNewMessages()
+}
+
+// EnqueueMessage adds a message to the log set but does not log it to the host.
+func (l *Logger) EnqueueMessage(severity diag.Severity, s string) {
```

Comment:
We're never calling `EnqueueMessage`, are we? If we're still calling `LogMessage` everywhere, we're flushing all these messages every time, are we not? So it seems like they should not be actually de-dup'd.

Reply:
Oops, good catch. I intended to enqueue the message as well.

Reply:
Oh, wait. I _am_ calling `EnqueueMessage` within `LogMessage`. That just bypasses the batching, but will still perform deduplication.

Reply:
Got it.